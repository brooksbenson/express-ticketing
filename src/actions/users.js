import db, { auth } from '../firebase/firebase';

export const addUser = ({ key, ...user }) => ({
  type: 'ADD_USER',
  key,
  user
});

export const startAddUser = user => {
  return async dispatch => {
    const { key } = await db.ref('user_data').push(user);
    dispatch(addUser({ key, ...user }));
    return key;
  };
};

export const setUsers = users => ({
  type: 'SET_USERS',
  users
});

export const startSetUsers = () => {
  return async dispatch => {
    const snap = await db.ref('user_data').once('value');
    dispatch(setUsers(snap.val()));
  };
};
