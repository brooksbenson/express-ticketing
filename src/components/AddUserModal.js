import React from 'react';
import Modal from 'react-modal';
import SearchBar from './SearchBar';
import userSelector from '../selectors/users';

Modal.setAppElement('body');

export default props => (
  <Modal
    className="add-user-modal"
    isOpen={props.isOpen}
    contentLabel="Add user modal"
    onRequestClose={props.onRequestClose}
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    closeTimeoutMS={200}
  >
    <h3> Add User </h3>
    <button onClick={props.onRequestClose}>
      <div />
      <div />
    </button>
    <SearchBar
      displayResults={!!props.searchString}
      onPick={props.onUserPick}
      onSearchChange={e => props.onSearchChange(e.target.value)}
      results={userSelector(props.users, props.searchString)}
      searchString={props.searchString}
    />
    <button onClick={props.onAddUser}>Add User</button>
  </Modal>
);
