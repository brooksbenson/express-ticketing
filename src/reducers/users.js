export default (state = {}, action) => {
  switch (action.type) {
    case 'ADD_USER':
      return { ...state, [action.key]: action.user };
    case 'SET_USERS':
      return action.users;
    default:
      return state;
  }
};
