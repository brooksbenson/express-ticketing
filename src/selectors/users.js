export default (users, search) => {
  const regex = new RegExp(search.split('').join('.*?'), 'i');
  return users.filter(user => regex.test(user.email));
};
