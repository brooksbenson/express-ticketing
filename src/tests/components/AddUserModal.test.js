import React from 'react';
import AddUserModal from '../../components/AddUserModal';
import { shallow } from 'enzyme';
import { usersArr } from '../fixtures/users';

let wrapper;
let onSearchChange;
let onAddUser;
let onRequestClose;
let onUserPick;

beforeEach(() => {
  onSearchChange = jest.fn();
  onAddUser = jest.fn();
  onUserPick = jest.fn();
  onRequestClose = jest.fn();
  wrapper = shallow(
    <AddUserModal
      isOpen={true}
      userPicked={false}
      users={usersArr}
      searchString=""
      onRequestClose={onRequestClose}
      onSearchChange={onSearchChange}
      onAddUser={onAddUser}
      onUserPick={onUserPick}
    />
  );
});

test('AddUserModal should match snapshot', () => {
  expect(wrapper).toMatchSnapshot();
});

test('AddUserModal should pass displayResults, searchString, onSearchChange, onPick, userPicked, and results to SearchBar component', () => {
  const searchBar = wrapper.find('SearchBar');
  expect(searchBar.prop('searchString')).toBe('');
  expect(searchBar.prop('onSearchChange')).toBeTruthy();
  expect(searchBar.prop('onPick')).toBe(onUserPick);
  expect(searchBar.prop('displayResults')).toBeFalsy();
  expect(searchBar.prop('results')).toEqual(usersArr);
});

test('AddUserModal should pass false to SearchBar prop displayResults when user has been picked', () => {
  wrapper.setProps({ searchString: 'Brooks Benson', userPicked: true });
  expect(wrapper.find('SearchBar').prop('displayResults')).toBeFalsy();
});

test('AddUserModal should pass true to SearchBar prop displayResults when user has not been picked', () => {
  wrapper.setProps({ searchString: 'Brooks', userPicked: false });
  expect(wrapper.find('SearchBar').prop('displayResults')).toBeTruthy();
});

test('AddUserModal should pass isOpen, onRequestClose, shouldCloseOnEsc, shouldCloseOnOverlayClick to Modal', () => {
  const modal = wrapper.find('Modal');
  expect(modal.prop('contentLabel')).toBe('Add user modal');
  expect(modal.prop('isOpen')).toBe(true);
  expect(modal.prop('onRequestClose')).toBe(onRequestClose);
  expect(modal.prop('shouldCloseOnEsc')).toBe(true);
  expect(modal.prop('shouldCloseOnOverlayClick')).toBe(true);
});

test('AddUserModal should invoke onRequestClose when close button is clicked', () => {
  wrapper
    .find('button')
    .at(0)
    .simulate('click');
  expect(onRequestClose).toHaveBeenCalled();
});

test('AddUserModal should invoke onAddUser on add user button click', () => {
  wrapper
    .find('button')
    .at(1)
    .simulate('click');
  expect(onAddUser).toHaveBeenCalled();
});
