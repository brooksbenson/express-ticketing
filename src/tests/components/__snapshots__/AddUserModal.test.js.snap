// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddUserModal should match snapshot 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="add-user-modal"
  closeTimeoutMS={200}
  contentLabel="Add user modal"
  isOpen={true}
  onRequestClose={[MockFunction]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <h3>
     Add User 
  </h3>
  <button
    onClick={[MockFunction]}
  >
    <div />
    <div />
  </button>
  <SearchBar
    displayResults={false}
    onPick={[MockFunction]}
    onSearchChange={[Function]}
    results={
      Array [
        Object {
          "admin": false,
          "email": "max@mail.com",
          "key": "kdlejf",
          "name": "Max Benson",
        },
        Object {
          "admin": true,
          "email": "brooks@mail.com",
          "key": "-keiajf",
          "name": "Brooks Benson",
        },
        Object {
          "admin": false,
          "email": "jeff@mail.com",
          "key": "-wqqadv",
          "name": "Jeff Benson",
        },
      ]
    }
    searchString=""
  />
  <button
    onClick={[MockFunction]}
  >
    Add User
  </button>
</Modal>
`;
